src/lib.rs

```diff
@@ -369,8 +369,15 @@ impl<W: Write> Archive<W> {
     }
 
     fn append_path2(&self, path: &Path) -> io::Result<()> {
-        let mut file = try!(fs::File::open(path));
-        self.append_file(path, &mut file)
+        let stat = try!(fs::metadata(path));
+        if stat.is_file() {
+            let mut file = try!(fs::File::open(path));
+            self.append_file(path, &mut file)
```

Comment:
Ah I was more going for the benefit of fewer syscalls rather than style here, `append_file` will call `metadata` again but we've already got the metadata here for the file. Currently the `foo2` functions are just the non-generic versions of `foo` to help codegen of downstream crates, so I wouldn't want to enhance the signatures too too much.